### ISSUE TITLE
Runtime of copy() for tables of different sizes varies hugely

### ISSUE THREAD

**Issue by awelzel** [zeek/zeek#2387]
Splitting this out from #2386: Runtime for copying a table of a simple string to count table varies wildly and unexpectedly. Initial population time and pure iteration time appear to increase reasonably, while `copy()` time increases 100 fold.

Running below script shows that the time required for copying() increases from ~1second for 1mio entries to ~50 seconds and 100seconds for 1.1mio and 1.2mio entries, respectively. At 1.7mio entries, the time is back at below 2 seconds.

```
$ time zeek ./large-copy.zeek 
Size, 1000000, fill, 0.713404, iterate, 0.453252, copy, 0.91069
Size, 1100000, fill, 1.137778, iterate, 0.492146, copy, 50.985543
Size, 1200000, fill, 1.3587, iterate, 0.514669, copy, 101.756562
Size, 1300000, fill, 1.444776, iterate, 0.577019, copy, 105.51632
Size, 1400000, fill, 1.581012, iterate, 0.608411, copy, 68.34778
Size, 1500000, fill, 1.683537, iterate, 0.670468, copy, 18.79153
Size, 1600000, fill, 1.990621, iterate, 0.709804, copy, 9.908489
Size, 1700000, fill, 2.000903, iterate, 0.736333, copy, 1.367809
Size, 1800000, fill, 2.257185, iterate, 0.804593, copy, 1.481103
Size, 1900000, fill, 2.200775, iterate, 1.085882, copy, 1.5915
Size, 2000000, fill, 2.303482, iterate, 0.908776, copy, 1.784987
Size, 2100000, fill, 2.320749, iterate, 1.087852, copy, 10.341843
```

```
module Test;

global tbl: table[string] of count;

function fill_table(limit: count)
        {
        local i = 0;
        while ( i < limit )
                {
                ++i;
                tbl[cat(i)] = i;
                }
        }

function iterate_table(t: table[string] of count)
        {
        local c = 0;
        for ( [k], v in t )
                c += v;
        }

event zeek_init()
        {
        local limit = 100000000;

        local one_million = 1000000;
        local i = one_million;
        local tbl2: table[string] of count;

        while ( i < limit )
                {
                tbl = table();
                local st = current_time();
                fill_table(i);
                local etf = current_time();
                iterate_table(tbl);
                local itf = current_time();
                tbl2 = copy(tbl);
                local etc = current_time();
                print "Size", i, "fill", interval_to_double(etf - st), "iterate", interval_to_double(itf - etf), "copy", interval_to_double(etc - itf);
                i = i + one_million / 10;
                }
        }
```

**Comment by timwoj** [comment]
One thing that might help here is setting the size of the dictionary when we create it in `TableVal::DoClone()`. That would avoid a large number of resizes. Alternatively, we could add a `Reserve()` or `Resize()` method to `Dictionary` and use that, similar to `List::resize()`. We already know what the size of the target dictionary should be, so we should just allocate that much space to start with.

That still doesn't explain why the time jumps around so much as the size grows though.

**Comment by timwoj** [comment]
I added some instrumentation to `DoClone()`. It's definitely something to do with the insertion into the new table. Even in the 1100000-element case above, eventually it hits a point where each insert takes a couple microseconds longer than the one before it, increasing forever.

On my local machine with `--build-type=debug`, I get the following:

```
long clones: 31
long inserts: 7809
Size, 1000000, fill, 5.244919, iterate, 1.824139, copy, 1.679316
long clones: 119
long inserts: 249487
Size, 1100000, fill, 5.782929, iterate, 2.012229, copy, 254.88545
```

Those two printfs keep track of the "long" operations in the loop in `DoClone()`. One of those is considered long if it takes more than 0.00001 seconds measured using `util::current_time()`. Most operations are well below that threshold (half of it or lower).

Digging a little further into `Dictionary::Insert`, it looks like we're spending the increasing time in `LookupIndex` and `InsertRelocateAndAdjust`. Both of these make sense as to why they're increasing time as the Dictionary grows, but not why it only does that when you hit specific table-size thresholds. I'd be curious to see exactly what the thresholds are that cause it to blow up.

**Comment by timwoj** [comment]
This is from Instruments.app, with sizes 1400000 and 1500000.

![Screen Shot 2022-08-30 at 11 47 15 AM](https://user-images.githubusercontent.com/2653616/187518422-986fbbe1-1d5b-4b99-8f59-8297ac07272f.png)
![Screen Shot 2022-08-30 at 11 47 29 AM](https://user-images.githubusercontent.com/2653616/187518443-63cc841f-a6ee-4cc6-afda-f95a4bcc83be.png)

Clearly remapping is causing a problem. I did some tests trying to find a fixed breakpoint where the time increases dramatically, but once you start filtering down into smaller sizes it just increases or decreases by small amounts between the runs. It's not that we're hitting a point where `Dictionary` suddenly performs badly, more that it just takes longer and longer and then reverses.

**Comment by timwoj** [comment]
Some more notes, just so I can keep this straight in my brain. The big question mark right now is why the initial insertion and the insertion during `copy()` is different. My current assumption is that because of the ordering of table elements, we're hitting some pathological case while looping on the existing order, and it's causing `Remap()` to take a very long time to do what it should. Forcing the dictionary to be ordered doesn't seem to change anything though so that assumption could be wildly wrong.

**Comment by awelzel** [comment]
For reproducing smaller sizes might be easier.  With 127k entries it takes 0.37sec with a asan/debug build. With 155k, it takes 3.8secs.

The flamegraph produced with those smaller sizes just shows `InsertAndRelocate()` and also `EndOfClusterByPosition()`. I'm also thinking that the iteration order is such that there's one run of elements that all end-up in the same bucket. I'm very familiar with this all, but `InsertAndRelocate()` reads as if it's replacing elements when there's a collision. `EndOfClusterByPosition()` looks like looping over elements in a bucket/cluster, too, which the flamegraph points out.

![155k-flamegraph](https://user-images.githubusercontent.com/2100509/187930970-ac4c597c-667a-4ca3-87da-4b62de1b6fed.svg)

**Comment by timwoj** [comment]
One thing that came to me last night (while I was doing the dishes) was a question of why we are doing this re-insertion thing anyways for `Dictionary`. We've already done all of the work to build the buckets and whatnot. Why does `Dictionary` not have a copy constructor, where it could just rebuild a new `Dictionary` with all of the information it already spent the time building once? It would need some smarts about how to copy the values stored in the `DictEntry` objects, maybe a callback method of some sort you could pass in, but that has to be considerably faster, right?

**Comment by awelzel** [comment]
Yeah, that's an interesting thought, but it would be for the C++ side of things only, right?

I had used the following Zeek implementation instead of the `copy()` bif (not tested, just typed out here):

```
function my_copy(t: table[string] of count): table[string] of count
    {
    local t2: table[string] of count;
    for ([k], c in t)
        t2[k] = c;
     return t2;
    }
```
It's showing the exact same increased runtimes as `copy()` (which I guess is good and bad). It would be nice if it wouldn't because practically it means there's an insert order for keys that triggers pathological cases.

**Comment by timwoj** [comment]
The problem is that you're re-inserting the keys after they've already been reordered by the first insertion, and ending up hitting the same pathological case as `TableVal::DoClone()`. It's why I was looking at testing with an ordered dictionary (and hence resulting in #2392).

> Yeah, that's an interesting thought, but it would be for the C++ side of things only, right?

`copy` in script-land ends up just calling `TableVal::DoClone()`. If `DoClone()` called a fast copy constructor instead of looping over all of the elements and reinserting them into a new dictionary, it should end up being faster.

**Comment by timwoj** [comment]
I added some code in 9ff8f3b079436a4759d6ceeae617374738f70552 that special cases a `Dictionary::Clone` method for `TableVal` that lets it just iterate over the table array and rebuild a new table without having to do new inserts. It's pretty wildly faster, and matches more what I'd expect from this kind of testing:

```
Size, 1000000, fill, 5.130216, iterate, 1.882342, copy, 0.369629
Size, 1100000, fill, 5.735257, iterate, 2.106349, copy, 0.564383
Size, 1200000, fill, 6.197535, iterate, 2.314797, copy, 0.585921
Size, 1300000, fill, 6.754121, iterate, 2.492996, copy, 0.652999
Size, 1400000, fill, 7.320729, iterate, 2.692462, copy, 0.715663
Size, 1500000, fill, 7.792278, iterate, 2.912998, copy, 0.775774
```

This is with a debug build so release would be faster, obviously. No giant spikes or valleys in time spent, just a steady increase as the table increases in size. There's some things wrong with it still, such as the fact that I didn't implement the cloning for `subnets`, and it could probably be made more generic where you pass a clone method for the `value` in `DictEntry` similar to how we do the deletions. As a PoC though, it's pretty obvious that this is the problem.

Unfortunately, it doesn't fix the root problem of the `Dictionary` having some glaring performance issues with some test cases so its kind of a band-aid.

**Comment by jasonlue** [comment]
I've created a unit test in Dict.cc to test the scenarios out. Here's what I found so far.
- It can be reproduced in C++ directly when inserting items in the order of their final location in the dictionary.
- When I randomize the items, the performance is normal. (This proves in the other way that the order is the issue)
- When the dictionary size is set to the final size (so that increase of dictionary size and remapping of old doesn't happen), the performance is normal.

More investigation is scheduled to understand the why part.

**Comment by JustinAzoff** [comment]
>  inserting items in the order of their final location in the dictionary.

Sounds like the sort of thing that happens when you insert sorted items into a binary tree and it degrades into a linked list.

**Comment by timwoj** [comment]
> When the dictionary size is set to the final size (so that increase of dictionary size and remapping of old doesn't happen), the performance is normal.

This happens independent of any other changes? We can certainly add some code to allow people to set the size of a dictionary when creating it or a method like `std::vector::reserve()`.

**Comment by jasonlue** [comment]
Yes. the only change of preset dictionary size  seems to bring the performance to normal.

My first guess is that Fibonacci Hashing mapping original hash to bucket mechanism caused the issue (It uses lower bits of the hash to indicate the bucket instead of common mod by prime).  Possible for some reason the hash doesn't distribute to all buckets evenly.  I added prime mod and it doesn't solve the problem.

The other guess is that the  dictionary is just too full and it's hard to land a empty space when inserting. The current load factor is 1/0.75, ie. grow the table when it's 75% full. I tried to make it 1/0.5 and it doesn't help much.

Justin's though was what I originally speculated. For some reason all items are concentrated on the front part of the dictionary, just like binary tree: with all the items are on the left pointer. 

I scheduled a few more days next week trying to understand why.

**Comment by jasonlue** [comment]
I post the unit test I am running here.  Put it in Dict.cc and run ./zeek --test -ts=dict -tc="dict clone"  to monitor the results by yourself.  FYI, use use stderr instead of stdout as stdout is blocked by doctest.

struct ReportDuration
	{
	std::chrono::_V2::system_clock::time_point start;
	ReportDuration()
		{
		start = std::chrono::high_resolution_clock::now();
		}
	~ReportDuration()
		{
		auto stop = std::chrono::high_resolution_clock::now();
		auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(stop - start);
		long d = duration.count();
		fprintf(stderr, "%ld ms\n", d);
		}
	};

TEST_CASE("dict clone")
	{
	for ( auto dictSize = 10000; dictSize < 2000000; dictSize += 10000 )
		{
		int* v = new int[dictSize];
		for ( int i = 0; i < dictSize; i++ )
			v[i] = i;
		PDict<int> dict;
		dict.Clear();
		fprintf(stderr, "%d \n", dictSize);
		fprintf(stderr, "fill:");
			{
			ReportDuration rd;
			for ( int i = 0; i < dictSize; i++ )
				{
				detail::HashKey* key = new detail::HashKey(v[i]);
				dict.Insert(key, &v[i]);
				}
			}
		int* v2 = new int[dictSize];
		int i = 0;
		for ( const auto& entry : dict )
			{
			auto val = static_cast<int*>(entry.value);
			v2[i++] = *val;
			}

		PDict<int> dictClone(UNORDERED, dictSize * 2);
		fprintf(stderr, "fill in final iteration order with preset size:");
			{
			ReportDuration rd;
			for ( int i = 0; i < dictSize; i++ )
				{
				detail::HashKey* key = new detail::HashKey(v2[i]);
				dictClone.Insert(key, &v2[i]);
				}
			}

		PDict<int> dictClone2;
		fprintf(stderr, "fill in final iteration order:");
			{
				{
				//				fprintf(stderr, "first half ");
				ReportDuration rd;
				for ( int i = 0; i < dictSize / 2; i++ )
					{
					detail::HashKey* key = new detail::HashKey(v2[i]);
					dictClone2.Insert(key, &v2[i]);
					}
				//				}
				//				{
				//				fprintf(stderr, "second half ");
				//				ReportDuration rd;
				for ( i = dictSize / 2; i < dictSize; i++ )
					{
					detail::HashKey* key = new detail::HashKey(v2[i]);
					dictClone2.Insert(key, &v2[i]);
					}
				}
			}
		PDict<int> dictClone3;
		std::random_shuffle(v2, v2 + dictSize);
		fprintf(stderr, "fill in random order:");
			{
			ReportDuration rd;
			for ( int i = 0; i < dictSize; i++ )
				{
				detail::HashKey* key = new detail::HashKey(v2[i]);
				dictClone3.Insert(key, &v2[i]);
				}
			}

		delete[] v;
		delete[] v2;
		}
	}

**Comment by timwoj** [comment]
I made some changes to the `TableVal` implementation to pass the size of the original table down to the new table before doing the clone, but it doesn't seem to have made any difference using the original test script. Patch is below.

```
diff --git a/src/Val.cc b/src/Val.cc
index 610df6204..ed1addd36 100644
--- a/src/Val.cc
+++ b/src/Val.cc
@@ -1387,9 +1387,9 @@ static void find_nested_record_types(const TypePtr& t, std::set<RecordType*>* fo
 		}
 	}
 
-TableVal::TableVal(TableTypePtr t, detail::AttributesPtr a) : Val(t)
+TableVal::TableVal(TableTypePtr t, detail::AttributesPtr a, size_t initial_size) : Val(t)
 	{
-	Init(std::move(t));
+	Init(std::move(t), initial_size);
 	SetAttrs(std::move(a));
 
 	if ( ! run_state::is_parsing )
@@ -1408,7 +1408,7 @@ TableVal::TableVal(TableTypePtr t, detail::AttributesPtr a) : Val(t)
 		}
 	}
 
-void TableVal::Init(TableTypePtr t)
+	void TableVal::Init(TableTypePtr t, size_t initial_size)
 	{
 	table_type = std::move(t);
 	expire_func = nullptr;
@@ -1423,7 +1423,7 @@ void TableVal::Init(TableTypePtr t)
 		subnets = nullptr;
 
 	table_hash = new detail::CompositeHash(table_type->GetIndices());
-	table_val = new PDict<TableEntryVal>;
+	table_val = new PDict<TableEntryVal>(UNORDERED, static_cast<int>(initial_size));
 	table_val->SetDeleteFunc(table_entry_val_delete_func);
 	}
 
@@ -2614,7 +2614,7 @@ double TableVal::CallExpireFunc(ListValPtr idx)
 
 ValPtr TableVal::DoClone(CloneState* state)
 	{
-	auto tv = make_intrusive<TableVal>(table_type);
+	auto tv = make_intrusive<TableVal>(table_type, attrs, Size());
 	state->NewClone(this, tv);
 
 	const PDict<TableEntryVal>* tbl = AsTable();
diff --git a/src/Val.h b/src/Val.h
index 19798422c..758e81802 100644
--- a/src/Val.h
+++ b/src/Val.h
@@ -8,6 +8,7 @@
 #include <unordered_map>
 #include <vector>
 
+#include "zeek/Dict.h"
 #include "zeek/IntrusivePtr.h"
 #include "zeek/Notifier.h"
 #include "zeek/Reporter.h"
@@ -737,7 +738,7 @@ protected:
 class TableVal final : public Val, public notifier::detail::Modifiable
 	{
 public:
-	explicit TableVal(TableTypePtr t, detail::AttributesPtr attrs = nullptr);
+	explicit TableVal(TableTypePtr t, detail::AttributesPtr attrs = nullptr, size_t initial_size = detail::DEFAULT_DICT_SIZE);
 
 	~TableVal() override;
 
@@ -998,7 +999,7 @@ public:
 	void EnableChangeNotifications() { in_change_func = false; }
 
 protected:
-	void Init(TableTypePtr t);
+	void Init(TableTypePtr t, size_t initial_size);
 
 	using TableRecordDependencies = std::unordered_map<RecordType*, std::vector<TableValPtr>>;
 
```

**Comment by jasonlue** [comment]
Hi, Tim,

This seems different from what I've tested with C++ unit test.  

I've developed a performance based table growth strategy to solve the problem and the initial results are very promising.
Along the way, I discovered an important optimization to boost large dictionary performance around 30% in performance. 

With the new strategy, you don't have to set the size explicitly, but if you do, the performance should be the best.

I plan to send a PR EOD Wednesday 10/5/2022.  At that time, You can run my unit test in my branch and your tests too to compare the results. I believe I have understood the problem and it should be solved.
